hyperstart/api/grpc/utils.go

```diff
@@ -0,0 +1,140 @@
+package grpc
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func copyValue(to, from reflect.Value) error {
+	toKind := to.Kind()
+	fromKind := from.Kind()
+
+	if !from.IsValid() {
+		return nil
+	}
+
+	if toKind == reflect.Ptr {
+		// If the destination is a pointer, we need to allocate a new one.
+		to.Set(reflect.New(to.Type().Elem()))
+		if fromKind == reflect.Ptr {
+			return copyValue(to.Elem(), from.Elem())
+		} else {
+			return copyValue(to.Elem(), from)
+		}
+	} else {
+		// Here the destination is not a pointer.
+		// Let's check what's the origin.
+		if fromKind == reflect.Ptr {
+			return copyValue(to, from.Elem())
+		}
+
+		switch toKind {
+		case reflect.Struct:
+			return copyStructValue(to, from)
+		case reflect.Slice:
+			return copySliceValue(to, from)
+		case reflect.Map:
+			return copyMapValue(to, from)
+		default:
+			// We now are copying non pointers scalar.
+			// This is the leaf of the recursion.
+			if from.Type() != to.Type() && from.Type().ConvertibleTo(to.Type()) {
+				to.Set(from.Convert(to.Type()))
+				return nil
+			} else {
+				to.Set(from)
+				return nil
+			}
+
+		}
+	}
+}
+
+func copyMapValue(to, from reflect.Value) error {
+	if to.Kind() != reflect.Map && from.Kind() != reflect.Map {
+		return fmt.Errorf("Can only copy maps into maps")
+	}
+
+	mapLen := from.Len()
+	to.Set(reflect.MakeMap(to.Type()))
+
+	for j := 0; j < mapLen; j++ {
+		if err := copyValue(to.Index(j), from.Index(j)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func copySliceValue(to, from reflect.Value) error {
+	if to.Kind() != reflect.Slice && from.Kind() != reflect.Slice {
+		return fmt.Errorf("Can only copy slices into slices")
+	}
+
+	sliceLen := from.Len()
+	to.Set(reflect.MakeSlice(to.Type(), sliceLen, sliceLen))
+
+	for j := 0; j < sliceLen; j++ {
+		if err := copyValue(to.Index(j), from.Index(j)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func copyStructValue(to, from reflect.Value) error {
+	if to.Kind() != reflect.Struct && from.Kind() != reflect.Struct {
+		return fmt.Errorf("Can only copy structs into structs")
+	}
+
+	if to.NumField() != from.NumField() {
+		return fmt.Errorf("Structures must have the same number of fields")
+	}
+
+	for i := 0; i < to.NumField(); i++ {
+		if err := copyValue(to.Field(i), from.Field(i)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func copyStruct(to interface{}, from interface{}) error {
+	toVal := reflect.ValueOf(to)
```

Comment:
You may want to consider trapping panics and returning an error here just in case there's an error case that isn't explicitly handled by the code.